spatialmath/capsule.go

```diff
@@ -0,0 +1,346 @@
+	segB r3.Vector // Distal endpoint of capsule line segment. Most distal point to be surrounded by `radius` of capsule
+}
+
+// NewCapsule instantiates a new capsule Geometry.
+func NewCapsule(offset Pose, radius, length float64, label string) (Geometry, error) {
+	if radius < 0 {
+		return nil, newBadGeometryDimensionsError(&capsule{})
+	}
+	if length == radius*2 || length == 0 {
+		return NewSphere(offset, radius, label)
+	}
+	if length < radius*2 {
+		return nil, newBadCapsuleLengthError(length, radius)
+	}
+	return newCapsuleWithSegPoints(offset, radius, length, label), nil
+}
+
+// Will precalculate the linear endpoints for a capsule.
+func newCapsuleWithSegPoints(offset Pose, radius, length float64, label string) Geometry {
+	segA := Compose(offset, NewPoseFromPoint(r3.Vector{0, 0, -length/2 + radius})).Point()
+	segB := Compose(offset, NewPoseFromPoint(r3.Vector{0, 0, length/2 - radius})).Point()
+
+	return &capsule{
+		pose:   offset,
+		radius: radius,
+		length: length,
+		label:  label,
+		segA:   segA,
+		segB:   segB,
+	}
+}
+
+func (c *capsule) MarshalJSON() ([]byte, error) {
+	config, err := NewGeometryConfig(c)
+	if err != nil {
+		return nil, err
+	}
+	config.Type = "capsule"
+	config.R = c.radius
+	config.L = c.length
+	return json.Marshal(config)
+}
+
+// String returns a human readable string that represents the capsule.
+func (c *capsule) String() string {
+	return fmt.Sprintf("Type: Capsule, Radius: %.0f, Length: %.0f", c.radius, c.length)
+}
+
+// Label returns the label of this capsule.
+func (c *capsule) Label() string {
+	if c != nil {
+		return c.label
+	}
+	return ""
+}
+
+// Pose returns the pose of the capsule.
+func (c *capsule) Pose() Pose {
+	return c.pose
+}
+
+// AlmostEqual compares the capsule with another geometry and checks if they are equivalent.
+func (c *capsule) AlmostEqual(g Geometry) bool {
+	other, ok := g.(*capsule)
+	if !ok {
+		return false
+	}
+	return PoseAlmostEqual(c.pose, other.pose) &&
+		utils.Float64AlmostEqual(c.radius, other.radius, 1e-8) &&
+		utils.Float64AlmostEqual(c.length, other.length, 1e-8)
+}
+
+// Transform premultiplies the capsule pose with a transform, allowing the capsule to be moved in space.
+func (c *capsule) Transform(toPremultiply Pose) Geometry {
+	return newCapsuleWithSegPoints(Compose(toPremultiply, c.pose), c.radius, c.length, c.label)
+}
+
+// ToProto converts the capsule to a Geometry proto message.
+func (c *capsule) ToProtobuf() *commonpb.Geometry {
+	return &commonpb.Geometry{
+		Center: PoseToProtobuf(c.pose),
+		GeometryType: &commonpb.Geometry_Capsule{
+			Capsule: &commonpb.Capsule{
+				RadiusMm: c.radius,
+				LengthMm: c.length,
+			},
+		},
+		Label: c.label,
+	}
+}
+
+// CollidesWith checks if the given capsule collides with the given geometry and returns true if it does.
+func (c *capsule) CollidesWith(g Geometry) (bool, error) {
+	if other, ok := g.(*box); ok {
+		return capsuleVsBoxCollision(c, other), nil
+	}
+	dist, err := c.DistanceFrom(g)
```

Comment:
For the other geometry types I found that it was not efficient to return the distance if we want simple collision, so I wrote two methods, one for each, for optimality's sake.  Do you think theres benefit to be had here or no?